0000-template.md

```diff
@@ -1,74 +1,89 @@
-- Start Date: (fill me in with today's date, YYYY-MM-DD)
-- RFC PR: (leave this empty)
-- Rust Issue: (leave this empty)
-
+- Feature Name: N/A
+- Start Date: June 1, 2018
+- RFC PR:
+- Rust Issue:
 # Summary
 [summary]: #summary
 
-One paragraph explanation of the feature.
+Allow non-ASCII identifiers, with strict checks to prevent confusion.
 
 # Motivation
 [motivation]: #motivation
 
-Why are we doing this? What use cases does it support? What is the expected outcome?
+One of the main use cases is Rust developers for whom English is not their native language.  This will allow them to use their native languages in identifiers.  This is very helpful, since identifiers often have significant semantic meaning.
 
 # Guide-level explanation
 [guide-level-explanation]: #guide-level-explanation
 
-Explain the proposal as if it was already included in the language and you were teaching it to another Rust programmer. That generally means:
+Rust supports a subset of Unicode in identifiers.  Therefore,
+
+```rust
+use num::complex::Complex;
+
+/// The tolerance
+const ε: f64 = 0.000001;
+
+/// Riemann zeta function
+fn ζ(x: Complex) -> Complex {
+   ...
+}
+```
 
-- Introducing new named concepts.
-- Explaining the feature largely in terms of examples.
-- Explaining how Rust programmers should *think* about the feature, and how it should impact the way they use Rust. It should explain the impact as concretely as possible.
-- If applicable, provide sample error messages, deprecation warnings, or migration guidance.
-- If applicable, describe the differences between teaching this to existing Rust programmers and new Rust programmers.
+There are restrictions on what can be put in identifiers, so the following is incorrect:
 
-For implementation-oriented RFCs (e.g. for compiler internals), this section should focus on how compiler contributors should think about the change, and give examples of its concrete impact. For policy RFCs, this section should provide an example-driven introduction to the policy, and explain its impact in concrete terms.
+```rust
+/// ERROR ☆ is not allowed in identifiers
+const ☆: f64 = 0;
+```
+
+Additionally, Rust includes checks to make sure that no two identifiers can be confused with each other.  Specifically:
+
+* It is a hard error if two identifiers can be confused with each other and are in scope at the same time.
+* It is a hard error if identifiers are not in NFC.
+* If two identifiers that can be confused with each other are present in the same source file, a warning is issued.
 
 # Reference-level explanation
 [reference-level-explanation]: #reference-level-explanation
 
-This is the technical portion of the RFC. Explain the design in sufficient detail that:
+This RFC proposes that non-ASCII identifiers be stabilized, with the current set of allowed characters.  However, there are restrictions to avoid visual confusion:
+
+* No two identifiers that are distinct but visually confusable may be in the same scope.
```

Comment:
How does this interact with `use foo::*; use bar::*;`?  Does this mean that minor version updates to crates can stop my code compiling because they introduce distinct-but-confusable names?